File: typescript/src/patterns/behavioral/state/Application.ts
import { IllegalStateError } from '../../../shared/errors.js';
import { JulLogger } from '../../JulLogger.js';
import { Order } from './Order.js';

const log = JulLogger.getLogger('pl.training.patterns.behavioral.state.Application');

export function run(): void {
  const order = new Order();
  log.info(`State: ${String(order.getState())}`, 'main');
  order.pay();
  log.info(`State: ${String(order.getState())}`, 'main');
  order.ship();
  log.info(`State: ${String(order.getState())}`, 'main');
  try {
    order.cancel();
  } catch (exception) {
    if (!(exception instanceof IllegalStateError)) {
      throw exception;
    }
    log.info(exception.message, 'main');
  }
}

File: csharp/src/Training.Patterns/Behavioral/State/Application.cs
namespace Training.Patterns.Behavioral.State;

public class Application
{
    private static readonly JulLogger log = JulLogger.GetLogger(typeof(Application));

    public static void Run()
    {
        var order = new Order();
        log.Info("State: " + order.State);
        order.Pay();
        log.Info("State: " + order.State);
        order.Ship();
        log.Info("State: " + order.State);
        try
        {
            order.Cancel();
        }
        catch (InvalidOperationException exception)
        {
            log.Info(exception.Message);
        }
    }
}
